// react-inline/lib/transformObjectExpressionIntoStyleSheetObject.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = transformObjectExpressionIntoStyleSheetObject;

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _vm = require('vm');

var _vm2 = _interopRequireDefault(_vm);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _recast = require('recast');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var n = _recast.types.namedTypes;
var isBlank = /^\s*$/;

function transformObjectExpressionIntoStyleSheetObject(expr, context) {
  (0, _assert2.default)(n.ObjectExpression.check(expr), 'must be a object expression');

  context = _vm2.default.createContext((0, _objectAssign2.default)({}, context));

  context.evaluate = function (node) {
    return _vm2.default.runInContext((0, _recast.print)(node).code, this);
  };

  var result = {};

  expr.properties.forEach(function (property) {
    processTopLevelProperty(property.key, property.value, result, context);
  });

  return result;
}

function processTopLevelProperty(key, value, result, context) {
  var name = keyToName(key);

  (0, _assert2.default)(n.ObjectExpression.check(value), 'top-level value must be a object expression');

  result[name] = {};

  processProperties(value.properties, result[name], context);
}

function processProperties(properties, result, context) {
  properties.forEach(function (property) {
    processProperty(property.key, property.value, result, context);
  });
}

function processProperty(key, value, result, context) {
  var name = keyToName(key);

  if (canEvaluate(value, context)) {
    var val = context.evaluate(value);

    (0, _assert2.default)(typeof val === 'string' || typeof val === 'number', 'value must be a string or number');

    if (typeof val === 'string') {
      (0, _assert2.default)(!isBlank.test(val), 'string value cannot be blank');
    }

    result[name] = val;
  } else if (n.ObjectExpression.check(value)) {
    result[name] = {};

    processProperties(value.properties, result[name], context);
  } else if (n.UnaryExpression.check(value) && value.prefix === true && value.operator === '-') {
    (0, _assert2.default)(n.Literal.check(value.argument), 'invalid unary argument type');

    result[name] = -value.argument.value;
  } else {
    (0, _assert2.default)(false, 'invalid value expression type');
  }
}

function keyToName(key) {
  (0, _assert2.default)(n.Identifier.check(key) || n.Literal.check(key) && typeof key.value === 'string', 'key must be a string or identifier');

  return key.name || key.value;
}

function canEvaluate(expr, context) {
  if (n.Literal.check(expr)) {
    return true;
  } else if (n.Identifier.check(expr) && context.hasOwnProperty(expr.name)) {
    return true;
  } else if (n.MemberExpression.check(expr)) {
    return n.Identifier.check(expr.property) && canEvaluate(expr.object, context);
  } else if (n.BinaryExpression.check(expr)) {
    return canEvaluate(expr.left, context) && canEvaluate(expr.right, context);
  }

  return false;
}
